### PR TITLE
Add ground atmosphere and day/night shading to Model and 3D Tiles

### LINKED ISSUE #11717: Visualize ground atmosphere for 3D TIles/Models
Splitting up #4196 into smaller tasks

This task is closely related to #11716, but it focuses on ground atmosphere rather than fog.

In `GlobeFS.glsl`, the globe shader handles both fog and ground atmosphere. In the spirit of `Model`'s pipeline stages, I envision this being implemented as a `GroundAtmospherePipelineStage` that's separate from `FogPipelineStage`

Also, see https://github.com/CesiumGS/cesium/issues/11681#issuecomment-1865080064 for the proposed `scene.groundAtmosphere` object that will make it easier to propagate settings via the FrameState and/or automatic uniforms.

One detail to note: Some details like computing the atmosphere color would be expensive to duplicate if both fog and ground atmosphere are needed for the same model, so it's best to avoid that.

### PR DESCRIPTION
**Pull request by ptrgags**
# Description

:warning: This PR is built on top of #11744, so make sure that gets merged first.

This PR adds ground atmosphere and day/night shading to the `AtmospherePipelineStage` for models and 3D Tiles. For world-scale datasets like P3DT, this will make it look a lot nicer!

| P3DT without ground atmosphere | P3DT with ground atmosphere |
|---|---|
| image | image |

## Issue number and link

Fixes #11717

## Testing plan

Sandcastles for testing:

- [Comparison with Globe](http://localhost:8080/Apps/Sandcastle/index.html#c=lVRNj9owEP0rLieQkKNqb3y1dGlpJbq7Eml7ycXEQ2KtY0e2A0qr/vdO4gTCh+jCzTNvZt6beSHWyjqyE7AHQ6ZEwZ48ghVFRn/WsX7Ui+v3o1aOCQUm6g3Jn0gRYmNQ8F1zeBHxK5gR2TJpYVilNszCipVgLlMOjME+o3ZM6N90a3T2SxvJm0B/gPC/g3GkIhUEZM45eUm10waYFNaJmDwsSCgk2EjFtQhXPcChCrZnwrUDYqxwsNQ6kXDa4WER+pJ+NcbvgNaqaG5EJpzYgaWM837T2rPx0+xriZNOijA0d5m2eQoGxi1wq5NzIIYO6UTqDZwD6uABwg5Nz3GsO67e0xoXUOREKGTPpL9RpJAZtaneY319CUQjBwqKbSTwTvTYkPJSsUzEK5GkTqhKQ7PQhU8cpbaQsMyBPj0/fcZGtYLzmZ3uVWZpdKH4vKuuRUZqzRSPmXUSqguEOsHzfSqc0wotuay6V0b0viL9OAU0Gh+Q6cyb83KdLZkGOq7N6O/apt4dc43zbrOo+ZOjgJuMjiSu6O6S+v+W7qP5RSc3iZ064b7ea/wK3qi/48H7ZjR+I63Nbk65aeAGTT68zcrrH0+rb8uvIRndZf321H6nl+OPsnvD3sS6UsKsqqp+H0WWa4OfsJF9SgMHWS7x38sGmwJLHY2tHYw9eBJ0Sydc7Ijg0yt/1ySWzFrMbAsp1+I3nmk2CRB/USo140j1eQdGsrKCpe9nKx+klE4CfF6vdFrLDTNnnf8B)
- [OWT Base Globe + atmosphere](http://localhost:8080/Apps/Sandcastle/index.html#c=zRj5T9tK+l8ZZfuD0wbHzkVCoG8phYIEpduEdqUGwcSeJCNsjzUzBlLE/77fHD5yvLRPT9pdIWLPzHff42YTXScEfWc8CtGYcI5pgj5gQdCniE0JSjl7pCEJ0XSJrvAz5pNkkgQsERI9UvJEODpCCXlCJ0TQLHa/6T1nUgv0+oQlEggSPqk10MskQWiuqB6gGY4EaUyS1/owJycCAoIcWbKuXg4Vt2azuuWKh+WxjJlIF4TDasGeAEnyzABb0ADHhGN3Fi3HzNGMw4xjSVlygLyGXhMhaWK3KhqcYC7hDSdtR4EhtN9refsDt+139/39dnuw3zD7e61+u9fe7wzcfr/V67YGXtee+IO+32l3O26/0+r2up7XUft1fco4JWATw/bFwIeUk+DXggBLz/VaLc/b9wb9QafTavdAqEZ+6rl9fwCiDjyv6/X9/qBRwevut1ugQL/f9X3P77bNUd2CZOnvsAa0QW+/O+h3gX+v1e9VWAN94NoG8v1Bu93prUrV3u92Bn67owWosn7VvwQLmszPssRawYpxurLt/uvm+OPX4/HFyd3F57vrm3EePpMkIhJJGhFBJKwlXxrLNpvo6vjfx1/R9fcx+n52hny3VYltBWKxIILwE6ayMIB+tD+OzbE74yy+YMmxgMVF6Pgtr9+CQMgdGONnGmfxKOCEJKMUB+SUc8YPUMcoqaREJsDdlNOYSvpIhIvD0LECKIhXFGAZLJBDFHLdEFe5wSLiRmzu3GuqKGI4BLvksh+gNy8a4/VeUzEpg/b29tBILiOi3sqkFXprJWerymoMmzBkBpkriii9xEl4wh4JP2GR0m1S4/Op8+YlUMvXH95tAxULv7po3dYh/Utv6+0DQ3RSe/OySvgVHR1BESBBx/Fdr47+QIpNq9tpoPynjpTOa2iahY0Hq/5JJiSL0WiBQ6hTK2YI9JE9WbVG5cQYIkvojPG4tER2J2lMiiVE0TIlRdTeGPAx7Llnl9fH4yIVHnGUkbz+5NYwvzOO5zEUBsN3TJ7BrfcG7JHRsDi/gmrqnNnFRZJmEs2EfjYQTRgsg5/xXcxCEl1hSTjFEUSneakbeoXUNAFkgmXGyUUIRrCE3GJPlK933jBHA+PeQAoFeY0AfzIeqlJKBJpmqpBjCDHJELQPzrL5IlqiH3t+A/m3hSVI0EYpE1Ql9veTCnMsJadABdKjct5EPbdNeoUMdIaciuhHqOfV87NCPy3qiMVELlS6RPSBoLMrJJaJXIDkQokYY9iknJN5FmEOggqVvSF6wpCgJaEZ5JzUmyAqFCXH77geeltRwf2J9mxg1IclokWB+oje6d+3Bt3T6Pr0N6iAHjF9BochMLcK+T2hdAeLh2BKLgAXsRkCtaCXzGaZIIVuYkFjaIXKAFNIGKHgdAKW1PP4cHPcI8XNWd9ubAK+RW3Xa2g1SnFfbXSbx32ZlXkBgsAZg4LAJiVcpQpOAuIm7Mkph4Fqft6kITBWMWI7AXIqxVEiEuFUkFDRHBHYAiWPkLNBGyxb8K5DRPmeZ4K6ysuFgmoTGIYY4wk1u2zy0LW2OnLY8s6ENAKr8n76CGl6SYWEI+5syr+hpTbBahX7QoMH5T5bv1SvIwmeRiQ/KKYfY44FuDlar2rV1qRkOjdAjhE6wMkjFlUvQdJgkAdfQ3hFeAnEQhZkqui4QAkEPY2IWoGRQvo4qSncfPTKZz4XpymBGr2gUeisEVTwa1tuEGEhPmMdGZPaFAcPIWfppLYFVHcxCEORGuEmtYQlZAdonmIaFk+hpUKJ2QE/ZVK1DgXt7QCLyEz+CujHpJYyqLWQtUSZXkxqt78lMsSPce+k1kmfdwkLpppDpVX9EDJbY/yj7am/HVhPCyqJjgiNkHKyF9GdEs3AsyP608D7vd0yQU8g/CtMKplYUcFOayqAz2UcQWCqoWdotqkGsBGsclE3hWOdNZX0iaEGquizaaS6wUpK1PMmYILZdgpVuk2OAlRBw4UQ/WKjIy9ilfYCRReqBhQRqJtb5qUzA1Uvu86vQ3UaseBB22IXRhGB928qHa243qiEhYxRl4lzQucLCeVtm0rusqjL6fP9L3jacAaOW0k9WxLF2LPmRHRUCjqpHerj94fQlXAID65e4V++/wJpTbhcHjZhAT+L4uSbmpAq202F1cwpTFm4fG+DyHAx/k0tvQtd+a3j3DmRX8oDp2xQBikiyVwuVBMqgVyzWUBCG0COjkvVw4fwOLR4Q/TuHa1XZ411UVYp/6C3wz+D1UpvF9wpSVTngQ3DvwOnGQuH2o5vXkrE19ye4ZZjzbsKoSx+v97Lt3MEFzeNT+Cpfb0jqGkCHeF8fHWp2tVm8mtm0GVhqPgLmVRU0EJW+Hlt/FnT0yP51fXN6PTu6vrb6UqXvblYvSCoEVpF0w9F1Yok9Vg+qZkvI+pTh73XwLU+ESSCS/zB5pCiMe09cu3aARVbX7LC4TpgSKbZfLRgTx9UWYeS9g06lu6M+svJBnx+rVunWL1jrGmhqKMRXIgtB2RYiP+BTnZ++dsqHacpjPD6Dvt/5xn9/F0VVm6uf1OV6vK/EGe3OqtG0MIDLGSkx+AxY9EU8yuSZI5OLJV66yDzeUQ+ZND0EhgqT3U7R6np53oCrzb4RsUCepBekOCBhJPklxPBXyoqr6bTrI/blls5qe9UpfxYqdQwHz4L8R1Ly8q3cpXA5VfOcJngmAaXqtGvyID+yIvdRwNScsuBddUb3Xy+vPh0PkYHvwf/+frz6XCXROq76yc9c5YUVkyzhrr1q+2GIWuN2qH2yPs88P5J45RxiTIeOa7blCQGR0kimtMMUCHYhSg7o24sxecWuShfw2qrNrPpAfLTZwQ3AQrXfTULD9ch9uCirO99Byh/29IZZZW0HdoPUD99Xoc9bFZVO4R7E6Lh0ZZP5UjfhOBklkWRGrgnNWixAL+Baj8F2mBWYAv//aXZdF33sAnL7ZjSpOQa5f8A)
- [OWT base globe again, but with FPS monitor](http://localhost:8080/Apps/Sandcastle/index.html#c=zRj5T9tK+l8ZZfuD0wbHzkVCoG8phYIEpduEdqUGwcSeJCNsjzUzBlLE/77fHD5yvLRPT9pdIWLPzHff42YTXScEfWc8CtGYcI5pgj5gQdCniE0JSjl7pCEJ0XSJrvAz5pNkkgQsERI9UvJEODpCCXlCJ0TQLHa/6T1nUgv0+oQlEggSPqk10MskQWiuqB6gGY4EaUyS1/owJycCAoIcWbKuXg4Vt2azuuWKh+WxjJlIF4TDasGeAEnyTAGvAIZkms1HcH7GcUzEF8JHBFiFJXiBEAAAx+4sWo6Zo+UMM44lZckB8hp6TYSkid2qKHyCuYQ3nLQdBYbQfq/l7Q/ctt/d9/fb7cF+w+zvtfrtXnu/M3D7/Vav2xp4XXviD/p+p93tuP1Oq9vrel5H7df1KeOUgAkN2xcDH1JOgl8LAiw912u1PG/fG/QHnU6r3QOhGvmp5/b9AYg68Lyu1/f7g0YFr7vfboEC/X7X9z2/2zZHdQuSpb/DGtAGvf3uoN8F/r1Wv1dhDfSBaxvI9wftdqe3KlV7v9sZ+O2OFqDK+lX/EixoMj/LEmsFK8bpyrb7r5vjj1+Pxxcndxef765vxnm0TZKISCRpRASRsJZ8aSzbbKKr438ff0XX38fo+9kZ8t1WJRUUiMWCCMJPmMrCAPrR/jg2x+6Ms/iCJccCFheh47e8fgsCIXdgjJ9pnMWjgBOSjFIckFPOGT9AHaOkkhKZfHBTTmMq6SMRLg5DxwqgIF5RgGWwQA5RyHVDXKUSi4gbsblzr6miiOEQ7JLLfoDevGiM13tNxWQY2tvbQyO5jIh6K3Nc6K2VFK8qqzFswpAZJLooovQSJ+EJeyT8hEVKt0mNz6fOm5dALV9/eLcNVCz86qJ1W4dqUXpbbx8YopPam5dVwq/o6AhqBgk6ju96dfQHUmxa3U4D5T91pHReQ9MsbDxY9U8yIVmMRgscQllbMUOgj+zJqjUqJ8YQWUJnjMelJbI7SWNSLCGKlikpovbGgI9hzz27vD4eF6nwiKOM5PUnt4b5nXE8j6EwGL5j8gxuvTdgj4yGxfkVFF/nzC4ukjSTaCb0s4FowmAZ/IzvYhaS6ApLwimOIDrNS93QK6SmCSATLDNOLlQNtYTcYk+Ur3feMEcD495ACgV5jQB/Mh6qUkoEmmaq7mMIMckQdBvOsvkiWqIfe34D+beFJUjQRikTVCX295MKcywlp0AF0qNy3kQ9t016hQx0hpyK6Eeo59Xzs0I/LeqIxUQuVLpE9IGgsysklolcgORCiRhj2KSck3kWYQ6CCpW9IXrCkKAloRnknNSbICoUJcfvuB56W1HB/Yn2bGDUhyWiRYH6iN7p37cG3dPo+vQ3qIAeMX0GhyEwtwr5PaF0B4uHYEouABexGQK1oJfMZpkghW5iQWNohcoAU0gYoeB0ApbU8/hwc9wjxc1Z325sAr5FbddraDVKcV9tdJvHfZmVeQGCwBmDgsAmJVylCk4C4ibsySlnh2p+3qQhMFYxYjsBcirFUSIS4VSQUNE084AAWGeDNli24F2HiPI9zwR1lZcLBdUmMMw8xhNq1NnkoWttdeSw5Z0JaQRW5f30EdL0kgoJR9zZlH9DS22C1Sr2hQYPyn22fqleRxI8jUh+UEw/xhwLcHO0XtWqrUnJdG6AHCN0gJNHLKpegqTBIA++hvCK8BKIhSzIVNFxgRIIehoRtQIjhfRxUquXs1qQj4guTlMCNXpBo9BZI6jg17bcIMJCfMY6Mia1KQ4eQs7SSW0LqO5iEIYiNcJNaglLyA7QPMU0LJ5CS4USswN+yqRqHQra2wEWkZn8FdCPSS1lUGsha4kyvZjUbn9LZIgf495JrZM+7xIWTDWHSqv6IWS2xvhH21N/O7CeFlQSHREaIeVkL6I7JZqBZ0f0p4H3e7tlgp5A+FeYVDKxooKd1lQAn8s4gsBUQ8/QbFMNYCNY5aJuCsc6ayrpE0MNVNFn00h1g5WUqOdNwASz7RSqdJscBaiChgsh+sVGR17EKu0Fii5UDSgiUDe3zEtnBqpedp1fh+o0YsGDtsUujCIC799UOlpxvVEJCxmjLhPnhM4XEsrbNpXcZVGX0+f7X/C04Qwct5J6tiSKsWfNieioFHRSO9TH7w+hK+EQHly9wr98/wXSmnC5PGzCAn4Wxck3NSFVtpsKq5lTmLJw+d4GkeFi/Jtaehe68lvHuXMiv5QHTtmgDFJEkrlcqCZUArlms4CENoAcHZeqhw/hcWjxhujdO1qvzhrroqxS/kFvh38Gq5XeLrhTkqjOAxuGfwdOMxYOtR3fvJSIr7k9wy3HmncVQln8fr2Xb+cILm4an8BT+3pHUNMEOsL5+OpStavN5NfMoMvCUPEXMqmooIWs8PPa+LOmp0fyq+ub0end1fW305Uue3OxekFQI7SKph+KqhVJ6rF8UjMfUtSXEXuvgWt9IkgEl/iDzSFFY9p75Nq1Ayq2vmSFw3XA4lPHB1XWoaR9g46lO6P+0LIBn1/r1ilW7xhrWijqaAQXYssBGRbif6CTnV/+tkrHaQojvL7D/t95Rj9/V4WVm+vfVKW6/C/E2a3OqhG08AALGekxeMxYNMX8iiSZoxNLpd46yHwekQ8ZNL0EhspT3c5Ravq5nsCrDb5RsYAepBckeCDhJPnlRPCXisqr6TTr47blVk7qO1Upv20qNcx30kJ8x9Ky8q1cJXD5UTRcJjimwaVq9CsyoD/yYvfRgJTccmBd9UY3ny8vPp2P0cHvwX++/nw63CWR+kz7Sc+cJYUV06yhbv3Iu2HIWqN2qD3yPg+8f9I4ZVyijEeO6zYlicFRkojmNANUCHYhys6oG0vxuUUuytew2qrNbHqA/PQZwU2AwnVfzcLDdYg9uCjre98Byt+2dEZZJW2H9gPUT5/XYQ+bVdUO4d6EaHi05cs60jchOJllUaQG7kkNWizAb6DaT4E2mBXYwn9/aTZd1z1swnI7pjQpuUb5Pw)
- [Camera Distance Comparison](http://localhost:8080/Apps/Sandcastle/index.html#c=lVbbbts4EP2VqdEHGTDoZL1PvmSbJm1cIE2DtXf3oQoaWhrbRCjSICmnbuB/X4q0Qlm+pPaTyZkzlzNHIyVSaANLhs+oYAACn+EKNcsz8q+7i+JG4s5XUhjKBKq40YKXWADoBAV+lSnes+QJVRemlGtsFaYJ1XhLV6h2TQaVsnG6ZZqxP5Opktl/UvF0cxE1rfu62YtFLNptuExTuJ9LIxVSzrRhCXSuYcw46lgkrglTHNDYLugzZaZMkFiEwRspZxy3I3Suxx4SFWk8B8R1RRaKZcywJWpC0zTahPbV+Gz6aWUzbYHs1aXJpF7MUWGvdJzKWd3RXr2aZ1xOsO7gLl9d6GvQuh+tpnM8jSwB+QKYsNVT7mcUC1sZ0XP5bPFuEtbb1kBQ0AnHtHIbApJ0JWjGkls2mxsmih42hF57Q2i1dBmvFkjuvt19soFcB/WcleiF5UbJXKSX1e5Kz1iMqEgTqg3HYgJjObPj+5gbI4WV5E0RvRCi1xVEyRyt0NImDC68OHfpLIvZuPacGP1cS9O7YNso73gVrn4IDRytKBSxp+9qUW+zdFqZn+XsaGHbSjgt9sg+Bb/Zf0WDp+XY6A1KmR3NclTAG2/46/ekPPrn7vbLzXAM3ZOkX47ac7qbPrTtH28uZ1c0Q0Wv7VaiIikmHIWOTtoBpTsvso4MVW4dBk74pprPNMW/qZgh+VlHfRLpW5hVBSMOZHL3+9OIvWlqgFUPCkRAFU1/pUaxn3UGEsceCQ69beQw7K4rW6f9R0WHZHRml2SeYlR56QV7FMJ9P//joQXVc6d2/vOh2aw0OPyhTfE2HRIj7xUmTDOr5E5zqy5pxW5HHz3GsRgUP/AqgDkWTMD7FxdmDdljAHIMszW5dvdsCtEQ+pWZN71yoOpqy4kbueDMxA0HWwPaZ2gbbYdyEDulaSFjJrxlO8ohTM75Cio59xJQCgxUMfkufH//EnqpUbhuwcZqa63bHh7DFHaja1dXF0J0e1zXyBUHyBW75Iq3mt0mWNQJFgcJFgcJ3p8zpeqpzvAODw5aoVgcpVicQrEPHfgV+/hdu5Vn37ZfhP0GXFIe7Sy+FpyfnZ01e41Wo6/NiuOF7xrgA8sW0q6YXPGIkLbBbMHtV51uT3K7Ug1JtPZ5APrtKrSfsiWwdLDnMxYSTrUeeBJH7Jd9fV3029Z/B8qlG823JSpOV4Xb/Pzi1l8SQvpte9yPNFLyCVW1yP8D)

# Author checklist

- [x] I have submitted a Contributor License Agreement
- [x] I have added my name to `CONTRIBUTORS.md`
- [x] I have updated `CHANGES.md` with a short summary of my change
- [x] I have added or updated unit tests to ensure consistent code coverage
- [x] I have update the inline documentation, and included code examples where relevant
- [x] I have performed a self-review of my code
- [x] Other to-dos:
  - [x] Determine why the globe appears black before I toggle dynamic lighting
  - [x] Determine the new condition for adding `AtmospherePipelineStage` to the model rendering pipeline
  - [x] Figure out better handling for normals (e.g. P3DT doesn't have them!)
  - [x] One of the diffuse terms looks like it's using the wrong light direction (always facing the camera). See if that's related to my normal workaround
  - [x] Add a `showGroundAtmosphere` setting. Though think about how this will work between `Atmosphere` and `Globe`. 
  - [x] Try more datasets, e.g. the One World Terrain globe, non-world-scale datasets to see how they look
  - [x] Check through Sandcastle gallery to see if examples look nice or if they need further tweaking.
  - [x] Examine performance of OWT Base Globe

### CODE REVIEW

**Review by ptrgags — COMMENTED** [review]
Some notes

File: packages/engine/Source/Shaders/Model/AtmosphereStageFS.glsl [hunk 1]
```diff
@@ -63,6 +63,63 @@ void applyFog(inout vec4 color, vec4 groundAtmosphereColor, vec3 lightDirection,
...
 }
 
+#line 1010000
+
+float fade(float x, vec2 range, vec2 clampRange) {
+    return clamp((x - range.y) / (range.y - range.x), clampRange.x, clampRange.y);
+}
+
+
+void applyGroundAtmosphere(inout vec4 color, vec4 groundAtmosphereColor, vec3 positionWC, vec3 atmosphereLightDirection, /*vec3 normalEC,*/ float cameraDist) {
```
[the remark this answers is not in the document]
Comment (ptrgags) [reply packages/engine/Source/Shaders/Model/AtmosphereStageFS.glsl]: ah the problem was the camera distance. It's not the distance from the camera to the fragment, it's the height of the camera relative to the center of the earth (pulled from the view matrix).

**Comment by ptrgags** [comment]
I tried the [OWT base globe](http://localhost:8080/Apps/Sandcastle/index.html#c=zRj5T9tK+l8ZZfuD0wbHzkVCoG8phYIEpduEdqUGwcSeJCNsjzUzBlLE/77fHD5yvLRPT9pdIWLPzHff42YTXScEfWc8CtGYcI5pgj5gQdCniE0JSjl7pCEJ0XSJrvAz5pNkkgQsERI9UvJEODpCCXlCJ0TQLHa/6T1nUgv0+oQlEggSPqk10MskQWiuqB6gGY4EaUyS1/owJycCAoIcWbKuXg4Vt2azuuWKh+WxjJlIF4TDasGeAEnyzABb0ADHhGN3Fi3HzNGMw4xjSVlygLyGXhMhaWK3KhqcYC7hDSdtR4EhtN9refsDt+139/39dnuw3zD7e61+u9fe7wzcfr/V67YGXtee+IO+32l3O26/0+r2up7XUft1fco4JWATw/bFwIeUk+DXggBLz/VaLc/b9wb9QafTavdAqEZ+6rl9fwCiDjyv6/X9/qBRwevut1ugQL/f9X3P77bNUd2CZOnvsAa0QW+/O+h3gX+v1e9VWAN94NoG8v1Bu93prUrV3u92Bn67owWosn7VvwQLmszPssRawYpxurLt/uvm+OPX4/HFyd3F57vrm3EePpMkIhJJGhFBJKwlXxrLNpvo6vjfx1/R9fcx+n52hny3VYltBWKxIILwE6ayMIB+tD+OzbE74yy+YMmxgMVF6Pgtr9+CQMgdGONnGmfxKOCEJKMUB+SUc8YPUMcoqaREJsDdlNOYSvpIhIvD0LECKIhXFGAZLJBDFHLdEFe5wSLiRmzu3GuqKGI4BLvksh+gNy8a4/VeUzEpg/b29tBILiOi3sqkFXprJWerymoMmzBkBpkriii9xEl4wh4JP2GR0m1S4/Op8+YlUMvXH95tAxULv7po3dYh/Utv6+0DQ3RSe/OySvgVHR1BESBBx/Fdr47+QIpNq9tpoPynjpTOa2iahY0Hq/5JJiSL0WiBQ6hTK2YI9JE9WbVG5cQYIkvojPG4tER2J2lMiiVE0TIlRdTeGPAx7Llnl9fH4yIVHnGUkbz+5NYwvzOO5zEUBsN3TJ7BrfcG7JHRsDi/gmrqnNnFRZJmEs2EfjYQTRgsg5/xXcxCEl1hSTjFEUSneakbeoXUNAFkgmXGyUUIRrCE3GJPlK933jBHA+PeQAoFeY0AfzIeqlJKBJpmqpBjCDHJELQPzrL5IlqiH3t+A/m3hSVI0EYpE1Ql9veTCnMsJadABdKjct5EPbdNeoUMdIaciuhHqOfV87NCPy3qiMVELlS6RPSBoLMrJJaJXIDkQokYY9iknJN5FmEOggqVvSF6wpCgJaEZ5JzUmyAqFCXH77geeltRwf2J9mxg1IclokWB+oje6d+3Bt3T6Pr0N6iAHjF9BochMLcK+T2hdAeLh2BKLgAXsRkCtaCXzGaZIIVuYkFjaIXKAFNIGKHgdAKW1PP4cHPcI8XNWd9ubAK+RW3Xa2g1SnFfbXSbx32ZlXkBgsAZg4LAJiVcpQpOAuIm7Mkph4Fqft6kITBWMWI7AXIqxVEiEuFUkFDRHBHYAiWPkLNBGyxb8K5DRPmeZ4K6ysuFgmoTGIYY4wk1u2zy0LW2OnLY8s6ENAKr8n76CGl6SYWEI+5syr+hpTbBahX7QoMH5T5bv1SvIwmeRiQ/KKYfY44FuDlar2rV1qRkOjdAjhE6wMkjFlUvQdJgkAdfQ3hFeAnEQhZkqui4QAkEPY2IWoGRQvo4qSncfPTKZz4XpymBGr2gUeisEVTwa1tuEGEhPmMdGZPaFAcPIWfppLYFVHcxCEORGuEmtYQlZAdonmIaFk+hpUKJ2QE/ZVK1DgXt7QCLyEz+CujHpJYyqLWQtUSZXkxqt78lMsSPce+k1kmfdwkLpppDpVX9EDJbY/yj7am/HVhPCyqJjgiNkHKyF9GdEs3AsyP608D7vd0yQU8g/CtMKplYUcFOayqAz2UcQWCqoWdotqkGsBGsclE3hWOdNZX0iaEGquizaaS6wUpK1PMmYILZdgpVuk2OAlRBw4UQ/WKjIy9ilfYCRReqBhQRqJtb5qUzA1Uvu86vQ3UaseBB22IXRhGB928qHa243qiEhYxRl4lzQucLCeVtm0rusqjL6fP9L3jacAaOW0k9WxLF2LPmRHRUCjqpHerj94fQlXAID65e4V++/wJpTbhcHjZhAT+L4uSbmpAq202F1cwpTFm4fG+DyHAx/k0tvQtd+a3j3DmRX8oDp2xQBikiyVwuVBMqgVyzWUBCG0COjkvVw4fwOLR4Q/TuHa1XZ411UVYp/6C3wz+D1UpvF9wpSVTngQ3DvwOnGQuH2o5vXkrE19ye4ZZjzbsKoSx+v97Lt3MEFzeNT+Cpfb0jqGkCHeF8fHWp2tVm8mtm0GVhqPgLmVRU0EJW+Hlt/FnT0yP51fXN6PTu6vrb6UqXvblYvSCoEVpF0w9F1Yok9Vg+qZkvI+pTh73XwLU+ESSCS/zB5pCiMe09cu3aARVbX7LC4TpgSKbZfLRgTx9UWYeS9g06lu6M+svJBnx+rVunWL1jrGmhqKMRXIgtB2RYiP+BTnZ++dsqHacpjPD6Dvt/5xn9/F0VVm6uf1OV6vK/EGe3OqtG0MIDLGSkx+AxY9EU8yuSZI5OLJV66yDzeUQ+ZND0EhgqT3U7R6np53oCrzb4RsUCepBekOCBhJPklxPBXyoqr6bTrI/blls5qe9UpfxYqdQwHz4L8R1Ly8q3cpXA5VfOcJngmAaXqtGvyID+yIvdRwNScsuBddUb3Xy+vPh0PkYHvwf/+frz6XCXROq76yc9c5YUVkyzhrr1q+2GIWuN2qH2yPs88P5J45RxiTIeOa7blCQGR0kimtMMUCHYhSg7o24sxecWuShfw2qrNrPpAfLTZwQ3AQrXfTULD9ch9uCirO99Byh/29IZZZW0HdoPUD99Xoc9bFZVO4R7E6Lh0ZZP5UjfhOBklkWRGrgnNWixAL+Baj8F2mBWYAv//aXZdF33sAnL7ZjSpOQa5f8A), working nicely with styles and custom shaders!

image
image

I am noticing the performance isn't the best for this example, but I'll have to see how much of that is due to the custom shader/styling and how much of that is the added overhead of atmosphere lighting.

File: packages/engine/Source/Scene/FrameState.js [hunk 2]
```diff
@@ -302,6 +303,9 @@ function FrameState(context, creditDisplay, jobScheduler) {
     mieScaleHeight: undefined,
```
[the remark this answers is not in the document]
Comment (ptrgags) [reply packages/engine/Source/Scene/FrameState.js]: No need now that FrameState just stores the `Atmosphere` object!

**Comment by ptrgags** [comment]
One of the specs in `Model` is failing, the render test with [fog in sunlight](http://localhost:8080/Specs/SpecRunner.html?spec=Scene%2FModel%2FModel%20fog%20renders%20a%20model%20in%20fog%20(sunlight)). 

I stepped through the code with `debugCanvasWidth=400`, one weird thing I notice is that the sunny day color seems wrong (I thought I picked a time that would make it orange like a sunset...), it's just a slightly different shade of blue from `DynamicAtmosphereLightingType.NONE` :thinking:. Meanwhile when I use `SCENE_LIGHT`, _then_ it becomes orange.

I checked the enum values in the uniform map and in the shader, they look correct. I also made sure the atmosphere object was propagated to the frame state. So not sure what's going on yet.

File: packages/engine/Source/Shaders/Model/AtmosphereStageFS.glsl [hunk 3]
```diff
@@ -63,49 +64,139 @@ void applyFog(inout vec4 color, vec4 groundAtmosphereColor, vec3 lightDirection,
...
+        // daytime side of the globe and 0 on the nighttime side of the globe.
+        //
+        // Use this to select the diffuse + ground atmosphere on the daytime side of the globe,
+        // and just the ground atmosphere (which is much darker) on the nighttime side.
+        float dayNightMask = clamp(dot(normalWC, atmosphereLightDirectionWC), 0.0, 1.0);
+        vec3 dayNightColor = mix(groundAtmosphereColor.rgb, diffuseAndGroundAtmosphere, dayNightMask);
+
+        // Fade in the day/night color in as the camera height increases towards space.
+        float nightFade = fade(cameraHeight, czm_atmosphereNightFadeRange, vec2(0.05, 1.0));
+        finalAtmosphereColor = mix(diffuseAndGroundAtmosphere, dayNightColor, nightFade);
```
Comment (ptrgags) [inline packages/engine/Source/Shaders/Model/AtmosphereStageFS.glsl:137]: in `GlobeFS.glsl`, `u_nightFadeDistance`, I found some of the variable names and usage a bit backwards to my understanding, so I rephrased it and documented the conventions I'm using. 

It's mathematically equivalent, I just thought this way would be a little easier to understand.

**Review by ptrgags — COMMENTED** [review]
Some notes

**Comment by ptrgags** [comment]
@jjhembd @ggetz oh yikes, it looks like the render fog in sunlight spec fails in `main` too. Since that's a render test, it doesn't run in CI, so it must have been one of my last commits in #11744 that was incorrect

**Comment by ptrgags** [comment]
I fixed it, basically the dark color was exactly black, I thought it was supposed to be slightly lighter than that, so I removed the one expectation.

**Comment by ptrgags** [comment]
@ggetz @jjhembd I only briefly looked at performance of the OWT base globe with the FPS monitor. I would recommend doing more thorough performance testing, but here's my observations so far:

* with the default view with atmosphere off entirely, the frame rate varies quite a bit from 51-60 FPS. Not sure why that would be for an idling tileset, camera and mouse.
* Zooming in (again with no atmosphere), the frame rate drops, presumably for rendering many tiles at once.
* With the default view, turning atmosphere on doesn't have a noticeable impact above the noise
* Zoomed in however, turning on the ground atmosphere does seem to cause the frame rate to drop a bit more.

**Comment by ggetz** [comment]
Thanks @ptrgags for all the work here! We are going to close this for now until we can dedicate someone to finishing up the work the testing an work you suggested here.

**Pull request closed without merging**